### ISSUE TITLE
Identation within a WHEN clause

### ISSUE THREAD

**Issue by Hendrikto** [sqlfluff/sqlfluff#7062]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

SQLFluff currently wants no indentation for `WHEN` clauses, making them harder to read. This issue is closely related to #7063.

I suggest adding either

* an `indented_when` setting, analogous to `indented_joins` and `indented_using_on`, controlling whether the contents of `WHEN` clauses should be indented, or
* a more generic setting like `indented_clauses`, which would generalize the `indented_XYZ` options.

I prefer the latter option, as I consider it cleaner and more general. Looking at the example below, SQLFluff also complains about the indentation within the `BEGIN … END` clause, which could be addressed with the same setting, instead of adding ad hoc `indented_XYZ` for even more cases.

We may even want to think about deprecating `indented_joins` and `indented_using_on` in favor of something like:

```ini
[sqlfluff:indentation]
indented_clauses = ["JOIN", "ON", "WHEN", …]
```

New cases could be added easily, like `indented_clauses = ["BEGIN"]`.

### Use case

Consider the following (SQLite) `CREATE TRIGGER` statement:

```sql
❯ cat create_trigger.sql 
CREATE TRIGGER trig
AFTER UPDATE ON tab1
WHEN
	old.a IS NOT new.a
	OR old.b IS NOT new.b
	OR old.c IS NOT new.c
BEGIN
	INSERT INTO tab2 (x) VALUES ('UPDATE');
END;
```

SQLFluff complains that the `WHEN` clause should not be indented:

```
❯ sqlfluff --version
sqlfluff, version 3.4.2
❯ sqlfluff lint --dialect=sqlite create_trigger.sql 
== [create_trigger.sql] FAIL                                                                                                  
L:   4 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   5 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   6 | P:   1 | LT02 | Line should not be indented. [layout.indent]
[…]
All Finished 📜 🎉!
```

It suggest the following layout:

```
❯ sqlfluff fix --dialect=sqlite create_trigger.sql 
==== finding fixable violations ====
== [create_trigger.sql] FAIL                                                                                                  
L:   4 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   5 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   6 | P:   1 | LT02 | Line should not be indented. [layout.indent]
[…]
== [create_trigger.sql] FIXED
4 fixable linting violations found
```
```sql
❯ cat create_trigger.sql                          
CREATE TRIGGER trig
AFTER UPDATE ON tab1
WHEN
old.a IS NOT new.a
OR old.b IS NOT new.b
OR old.c IS NOT new.c
BEGIN
INSERT INTO tab2 (x) VALUES ('UPDATE');
END;
```

I consider the suggested layout much less readable than the one with indentation. I also think including indentation aligns more closely with most style guides.

### Dialect

any dialect

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by saulotoledo** [comment]
@Hendrikto, the issue here is that the parser does not recognize `IS NOT` in this context, causing the whole parsing to fail. If you replace `IS NOT` with `<>`, it will work.

Additionally, Oracle requires parentheses around the `WHEN` clause condition. The syntax should be `WHEN (condition)`, not `WHEN condition`.

In general, `IS NOT` is only valid with `NULL` comparisons (`IS NOT NULL`), not for general value comparisons. For value comparisons, use standard operators like `<>` or `!=`. `<>` is ANSI-compatible and works across Oracle, PostgreSQL, SQL Server, and other dialects.

ANSI does not support `CREATE TRIGGER`, so this is likely Oracle dialect. For Oracle triggers, ensure you use:

- `<>` or `!=` for inequality comparisons
- Parentheses around `WHEN` conditions: `WHEN (condition)`

If you think this should be supported by your dialect, could you please specify which one?


**EDIT:** there are some mistakes in my answer above:
-  `IS NOT` is not recognized in some parsers, but it does work on SQLite
- SQLite does not need parenthesis on the `WHEN` clause
- ANSI **does** support `CREATE TRIGGER`

**Comment by Hendrikto** [comment]
@saulotoledo thank you for your reply.

> […] causing the whole parsing to fail

I do not think the parsing fails entirely, as (a) I get no parsing error, (b) other problems are reported correctly, and (c) `sqlfluff fix` still works.

When parsing fails completely, you get a PRS error:

```sql
❯ cat parse_error.sql 
CREATE TRIGER trig
AFTER UPDATE ON tab1
BEGIN
	INSERT INTO tab2 (x) VALUES ('UPDATE');
END;
```
```
❯ sqlfluff lint --dialect=sqlite parse_error.sql
== [parse_error.sql] FAIL                                                                                                     
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'CREATE
                       | TRIGER trig\nAFTER UPDATE ON tab1\n...'
WARNING: Parsing errors found and dialect is set to 'sqlite'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

 > Additionally, Oracle requires parentheses around the WHEN clause condition. The syntax should be WHEN (condition), not WHEN condition.

SQLite does not: https://sqlite.org/lang_createtrigger.html

> If you replace IS NOT with <>, it will work.

First of all, it does not:

```sql
❯ cat create_trigger.sql                           
CREATE TRIGGER trig
AFTER UPDATE ON tab1
WHEN
	old.a <> new.a
	OR old.b <> new.b
	OR old.c <> new.c
BEGIN
	INSERT INTO tab2 (x) VALUES ('UPDATE');
END;
```
```
❯ sqlfluff lint --dialect=sqlite create_trigger.sql
== [create_trigger.sql] FAIL                                                                                                  
L:   4 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   5 | P:   1 | LT02 | Line should not be indented. [layout.indent]
L:   6 | P:   1 | LT02 | Line should not be indented. [layout.indent]
[…]
All Finished 📜 🎉!
```

Moreover, changing the operator, even if it worked, would also affect the semantics (see https://www.sqlite.org/lang_expr.html#operators_and_parse_affecting_attributes), and I actually do need `IS NOT` for my use case.

> If you think this should be supported by your dialect, could you please specify which one?

I do mention that I am using SQLite at the top of the "Use case" section:

> Consider the following (SQLite) CREATE TRIGGER statement:

But I think this should work for any dialect, and is not really specific to SQLite, although the specific syntax here is SQLite.

**Comment by saulotoledo** [comment]
Hi again, @Hendrikto.

Thanks again for the details. I will check the SQLite parser. In fact, there might be a bug with unsupported tokens that prevents parsing errors from being displayed properly.

We still need to check each dialect individually. For example, ANSI does not support `CREATE TRIGGER`, so it should fail there. Oracle does not support the `WHEN` clause without brackets. Other dialects might behave differently as well.

We might need help from someone who has worked with the SQLite parser before, but let me take a look at that part of the code.


**EDIT:** there are some mistakes in my answer above:
-  `IS NOT` is not recognized in some parsers, but it does work on SQLite
- ANSI **does** support `CREATE TRIGGER`

**Comment by saulotoledo** [comment]
@Hendrikto, you were correct. The parser for SQLite was working and it already supports `IS NOT`. There was indeed something missing in the indentation settings. I pushed a fix above.

**Comment by saulotoledo** [comment]
@Hendrikto, I was too fast when answering yesterday and missed the details, we indeed have `CREATE TRIGGER` for ANSI (I will correct my answer above) and we might miss indentation in other cases. Seeing the code, we still need to check each dialect though.

**Comment by saulotoledo** [comment]
@WittierDinosaur, this was auto-closed via the PR reference, but we still need to review the other dialects. I suggest reopening it—let us know if you see it differently.
